Fix imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.crossOriginSource.sub.html

youennf opened the pull request:
#### e428405c2a7e87e904e09a8122b3ae9d223b59aa
<pre>
Fix imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.crossOriginSource.sub.html
<a href="https://bugs.webkit.org/show_bug.cgi?id=257749">https://bugs.webkit.org/show_bug.cgi?id=257749</a>
rdar://110322419

Reviewed by Eric Carlson.

Update imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.crossOriginSource.sub.html to work around specificities related to not rendered SVG images.
The specificities are that the load event is not fired when the SVG image is detached from the DOM and that there is no CachedImage if it is not visible.
We should investigate both specificities independently.
Update SVGImageElement::renderingTaintsOrigin to take into account CORS and add a CORS check in the SVG Image to VideoFrame code path.
We move the data URL check after the imge renderingTaintsOrigin check as we should first check whether its content is not tainting and then if its origin would taint or not.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.crossOriginSource.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.crossOriginSource.sub.html:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::checkImageUsability):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::originClean const):
* Source/WebCore/svg/SVGImageElement.cpp:
(WebCore::SVGImageElement::renderingTaintsOrigin const):

Canonical link: <a href="https://commits.webkit.org/265142@main">https://commits.webkit.org/265142@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af283c425250ae2b54f813807f4dad9f739e087a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9712 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9963 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10208 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11372 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9484 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9723 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11949 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9922 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12408 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9866 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10706 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8266 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11532 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8010 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8829 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16233 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9108 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8977 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12311 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9471 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7817 "4 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8651 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8725 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2390 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12877 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9240 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
```diff
@@ -93,6 +93,9 @@ static std::optional<Exception> checkImageUsability(ScriptExecutionContext& cont
         return { };
     },
     [] (const RefPtr<SVGImageElement>& imageElement) -> std::optional<Exception> {
+        if (imageElement->renderingTaintsOrigin())
+            return Exception { SecurityError, "Image element is tainted"_s };
+
```
I think this statement should be moved after checking the `imageElement->cachedImage()` below. `SVGImageElement::renderingTaintsOrigin()` will return false if cachedImage is nullptr.

youennf replied:
Right, checkImageUsability should be renamed checkImageUsabilityAndOriginIsClean.
I'll do that in a follow-up.

youennf replied:
https://bugs.webkit.org/show_bug.cgi?id=258064

shallawa commented on Source/WebCore/svg/SVGImageElement.cpp:
```diff
@@ -86,8 +86,17 @@ bool SVGImageElement::renderingTaintsOrigin() const
 
-    auto* image = cachedImage->image();
-    return image && image->renderingTaintsOrigin();
+    RefPtr image = cachedImage->image();
+    if (!image)
+        return false;
+
+    if (image->renderingTaintsOrigin())
+        return true;
+
+    if (image->sourceURL().protocolIsData())
+        return false;
+
+    return cachedImage->isCORSCrossOrigin();
```
This part is almost the same the code in `HTMLImageElement::originClean`. Can we have a helper function in `CachedImage` so code is not duplicated?

youennf replied:
The issue is that there is a slight difference in when protocolIsData is checked: we check in first for SVG and last for images.

We should probably align with the SVG one, but it might be better to do it in a follow-up.
I'll file a bug to follow-up on this.

youennf replied:
https://bugs.webkit.org/show_bug.cgi?id=258065

youennf committed:
Fix imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.crossOriginSource.sub.html
https://bugs.webkit.org/show_bug.cgi?id=257749
rdar://110322419

Reviewed by Eric Carlson.

Update imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.crossOriginSource.sub.html to work around specificities related to not rendered SVG images.
The specificities are that the load event is not fired when the SVG image is detached from the DOM and that there is no CachedImage if it is not visible.
We should investigate both specificities independently.
Update SVGImageElement::renderingTaintsOrigin to take into account CORS and add a CORS check in the SVG Image to VideoFrame code path.
We move the data URL check after the imge renderingTaintsOrigin check as we should first check whether its content is not tainting and then if its origin would taint or not.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.crossOriginSource.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-construction.crossOriginSource.sub.html:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::checkImageUsability):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::originClean const):
* Source/WebCore/svg/SVGImageElement.cpp:
(WebCore::SVGImageElement::renderingTaintsOrigin const):

Canonical link: https://commits.webkit.org/265142@main
